include/picolibrary/circular_buffer.h

```diff
@@ -603,6 +604,464 @@ class Circular_Buffer<T, N, Without_Interrupt_Support, With_Overflow_Underflow_P
+     *
+     * \return Success.
+     */
+    auto push( Value && value ) noexcept -> Result<Void, Void>
+    {
+        auto guard = Interrupt_Guard<Restore_Interrupt_Enable_State>{};
+
+        new ( m_write ) Value{ std::move( value ) };
+
+        m_write = m_write + 1 == &m_storage[ N ] ? &m_storage[ 0 ] : m_write + 1;
+
+        ++m_size;
+
+        return {};
+    }
+
+    /**
+     * \brief Push a value to the end of the circular buffer.
+     *
+     * \warning Calling this function on a full circular buffer results in undefined
+     *          behavior.
+     *
+     * \param[in] value The value to push to the end of the circular buffer.
+     *
+     * \return Success.
+     */
+    auto push_from_interrupt( Value && value ) noexcept -> Result<Void, Void>
+    {
+        new ( m_write ) Value{ std::move( value ) };
+
+        m_write = m_write + 1 == &m_storage[ N ] ? &m_storage[ 0 ] : m_write + 1;
+
+        ++m_size;
+
+        return {};
+    }
+
+    /**
+     * \brief Push a value to the end of the circular buffer.
+     *
+     * \warning Calling this function on a full circular buffer results in undefined
+     *          behavior.
+     *
+     * \param[in] value The value to push to the end of the circular buffer.
+     *
+     * \return Success.
+     */
+    auto push( Value const & value ) noexcept -> Result<Void, Void>
+    {
+        auto guard = Interrupt_Guard<Restore_Interrupt_Enable_State>{};
+
+        new ( m_write ) Value{ value };
+
+        m_write = m_write + 1 == &m_storage[ N ] ? &m_storage[ 0 ] : m_write + 1;
+
+        ++m_size;
+
+        return {};
+    }
+
+    /**
+     * \brief Push a value to the end of the circular buffer.
+     *
+     * \warning Calling this function on a full circular buffer results in undefined
+     *          behavior.
+     *
+     * \param[in] value The value to push to the end of the circular buffer.
+     *
+     * \return Success.
+     */
+    auto push_from_interrupt( Value const & value ) noexcept -> Result<Void, Void>
+    {
+        new ( m_write ) Value{ value };
+
+        m_write = m_write + 1 == &m_storage[ N ] ? &m_storage[ 0 ] : m_write + 1;
+
+        ++m_size;
+
+        return {};
+    }
+
+    /**
+     * \brief Push a value to the end of the circular buffer, constructing the element
+     *        in-place.
+     *
+     * \warning Calling this function on a full circular buffer results in undefined
+     *          behavior.
+     *
+     * \tparam Arguments Value constructor argument types.
+     *
+     * \param[in] arguments Value constructor arguments.
+     *
+     * \return Success.
+     */
+    template<typename... Arguments>
+    auto emplace( Arguments &&... arguments ) noexcept -> Result<Void, Void>
+    {
+        auto guard = Interrupt_Guard<Restore_Interrupt_Enable_State>{};
+
+        new ( m_write ) Value{ std::forward<Arguments>( arguments )... };
+
+        m_write = m_write + 1 == &m_storage[ N ] ? &m_storage[ 0 ] : m_write + 1;
```

Comment:
Making a non-volatile copy of `m_write` and then using the copy would allow for performance optimizations.
```suggestion
        auto write = Storage volatile * { m_write };

        new ( write ) Value{ std::forward<Arguments>( arguments )... };

        m_write = write + 1 == &m_storage[ N ] ? &m_storage[ 0 ] : write + 1;
```